# Sample Data exploration

## Load Json data file

In [10]:
import json
import pandas as pd
import numpy as np

# Read JSON file
with open("data/train.json", "r", encoding="utf-8") as f:
    data = json.load(f)

# Check type
print(type(data))
print(len(data))

# First record
print(data[0])

# Convert to DataFrame
df = pd.DataFrame(data)

# View first rows
df.head()



<class 'list'>
2400
{'question': "Analyze the 5G wireless network drive-test user plane data and engineering parameters.\nIdentify the reason for the throughput dropping below 600Mbps in certain road sections.\nFrom the following 8 potential root causes, select the most likely one and enclose its number in \\boxed{{}} in the final answer.\n\nC1: The serving cell's downtilt angle is too large, causing weak coverage at the far end.\nC2: The serving cell's coverage distance exceeds 1km, resulting in over-shooting.\nC3: A neighboring cell provides higher throughput.\nC4: Non-colocated co-frequency neighboring cells cause severe overlapping coverage.\nC5: Frequent handovers degrade performance.\nC6: Neighbor cell and serving cell have the same PCI mod 30, leading to interference.\nC7: Test vehicle speed exceeds 40km/h, impacting user throughput.\nC8: Average scheduled RBs are below 160, affecting throughput.\n\nGiven:\n- The default electronic downtilt value is 255, representing a downtilt 

,question,answer
0,Analyze the 5G wireless network drive-test use...,C2
1,Analyze the 5G wireless network drive-test use...,C1
2,Analyze the 5G wireless network drive-test use...,C2
3,Analyze the 5G wireless network drive-test use...,C2
4,Analyze the 5G wireless network drive-test use...,C5


In [11]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 2400 entries, 0 to 2399
Data columns (total 2 columns):
 #   Column    Non-Null Count  Dtype 
---  ------    --------------  ----- 
 0   question  2400 non-null   object
 1   answer    2400 non-null   object
dtypes: object(2)
memory usage: 37.6+ KB


In [12]:
df.shape

(2400, 2)

Thre are 2400 rows

"Analyze the 5G wireless network drive-test user plane data and engineering parameters.\n
Identify the reason for the throughput dropping below 600Mbps in certain road sections.\n
From the following 8 potential root causes, select the most likely one and enclose its number in \\boxed{{}} in the final answer.\n\n
C1: The serving cell's downtilt angle is too large, causing weak coverage at the far end.\n
C2: The serving cell's coverage distance exceeds 1km, resulting in over-shooting.\n
C3: A neighboring cell provides higher throughput.\n
C4: Non-colocated co-frequency neighboring cells cause severe overlapping coverage.\n
C5: Frequent handovers degrade performance.\n
C6: Neighbor cell and serving cell have the same PCI mod 30, leading to interference.\n
C7: Test vehicle speed exceeds 40km/h, impacting user throughput.\n
C8: Average scheduled RBs are below 160, affecting throughput.\n\n

Given:\n- The default electronic downtilt value is 255, representing a downtilt angle of 6 degrees. Other values represent the actual downtilt angle in degrees.\n\nBeam 
Scenario and Vertical Beamwidth Relationships:\n- 
When the cell's Beam Scenario is set to Default or SCENARIO_1 to SCENARIO_5, the vertical beamwidth is 6 degrees.\n- 
When the cell's Beam Scenario is set to SCENARIO_6 to SCENARIO_11, the vertical beamwidth is 12 degrees.\n- 
When the cell's Beam Scenario is set to SCENARIO_12 or above, the vertical beamwidth is 25 degrees.\n\n

User plane drive test data as follows：\n\n

Timestamp|Longitude|Latitude|GPS Speed (km/h)|5G KPI PCell RF Serving PCI|5G KPI PCell RF Serving SS-RSRP [dBm]|5G KPI PCell RF Serving SS-SINR [dB]|5G KPI PCell Layer2 MAC DL Throughput [Mbps]|Measurement PCell Neighbor Cell Top Set(Cell Level) Top 1 PCI|Measurement PCell Neighbor Cell Top Set(Cell Level) Top 2 PCI|Measurement PCell Neighbor Cell Top Set(Cell Level) Top 3 PCI|Measurement PCell Neighbor Cell Top Set(Cell Level) Top 4 PCI|Measurement PCell Neighbor Cell Top Set(Cell Level) Top 5 PCI|Measurement PCell Neighbor Cell Top Set(Cell Level) Top 1 Filtered Tx BRSRP [dBm]|Measurement PCell Neighbor Cell Top Set(Cell Level) Top 2 Filtered Tx BRSRP [dBm]|Measurement PCell Neighbor Cell Top Set(Cell Level) Top 3 Filtered Tx BRSRP [dBm]|Measurement PCell Neighbor Cell Top Set(Cell Level) Top 4 Filtered Tx BRSRP [dBm]|Measurement PCell Neighbor Cell Top Set(Cell Level) Top 5 Filtered Tx BRSRP [dBm]|5G KPI PCell Layer1 DL RB Num (Including 0)\n


2025-05-07 15:23:52.000000|128.188169|32.579273|1|712|-77.0|15.93|1351.25|258|71|284|-|-|-90.49|-95.82|-102.2|-|-|210.91\n
2025-05-07 15:23:53.000000|128.18814|32.579223|2|71|-80.97|6.6|366.57|258|712|129|-|-|-88.72|-93.26|-105.47|-|-|211.23\n
2025-05-07 15:23:54.000000|128.188117|32.579174|16|71|-85.5|1.81|334.0|258|712|-|-|-|-84.41|-96.46|-|-|-|210.99\n
2025-05-07 15:23:55.000000|128.188103|32.579113|14|71|-88.21|5.4|431.94|712|258|-|-|-|-88.04|-93.72|-|-|-|195.26\n
2025-05-07 15:23:56.000000|128.188088|32.579075|19|71|-78.45|13.59|566.34|712|258|284|-|-|-87.97|-98.7|-105.92|-|-|209.0\n
2025-05-07 15:23:57.000000|128.188073|32.579033|22|258|-77.45|11.45|1308.74|712|71|284|129|-|-89.56|-93.31|-104.18|-110.78|-|204.95\n
2025-05-07 15:23:58.000000|128.188051|32.579002|16|258|-81.66|10.06|1310.14|712|71|284|129|-|-84.85|-93.3|-102.87|-114.09|-|210.13\n
2025-05-07 15:23:59.000000|128.188073|32.578995|26|258|-79.13|16.32|1023.36|712|71|284|129|-|-86.75|-93.71|-104.01|-107.26|-|208.06\n
2025-05-07 15:24:00.000000|128.188036|32.578984|33|258|-74.09|19.8|770.83|712|71|284|129|-|-89.29|-96.4|-100.21|-114.11|-|207.07\n
2025-05-07 15:24:01.000000|128.188036|32.578984|25|258|-71.09|19.62|1014.75|712|71|-|-|-|-90.14|-92.93|-|-|-|205.57\n\n\n

Engeneering parameters data as follows：\n\n

gNodeB ID|Cell ID|Longitude|Latitude|Mechanical Azimuth|Mechanical Downtilt|Digital Tilt|Digital Azimuth|Beam Scenario|Height|PCI|TxRx Mode|Max Transmit Power|Antenna Model\n
0034038|4|128.188273|32.579129|50|6|4|0|DEFAULT|5.0|712|64T64R|34.9|NR AAU 2\n0033164|27|128.217634|32.58072|240|3|10|0|DEFAULT|29.7|71|32T32R|34.9|NR AAU 1\n

0034038|2|128.188282|32.579099|140|10|4|0|DEFAULT|5.0|258|64T64R|34.9|NR AAU 2\n
0034038|24|128.18769|32.57791|40|44|9|10|SCENARIO_9|56.0|284|64T64R|34.9|NR AAU 2\n0033164|14|128.189481|32.581043|150|20|255|0|DEFAULT|28.7|129|64T64R|34.9|NR AAU 2\n",

## Data Wrangling
### Parse Tabular Table

In [ ]:
import re
import pandas as pd
from io import StringIO

# Function to parse the question text and extract the tables
def parse_question_tables(question_text):
    # Extract user plane table
    user_match = re.search(
        r"User plane drive test data as follows：\n\n(.*?)\n\n\nEngeneering parameters data as follows：",
        question_text,
        re.S
    )

    # Extract engineering table
    eng_match = re.search(
        r"Engeneering parameters data as follows：\n\n(.*)",
        question_text,
        re.S
    )

    user_table_text = user_match.group(1).strip()
    eng_table_text = eng_match.group(1).strip()

    user_df = pd.read_csv(StringIO(user_table_text), sep="|")
    eng_df = pd.read_csv(StringIO(eng_table_text), sep="|")

    return user_df, eng_df

### User Plane Table Explanation

In [7]:
user_df, eng_df = parse_question_tables(df.loc[0, "question"])

user_df.head()

,Timestamp,Longitude,Latitude,GPS Speed (km/h),5G KPI PCell RF Serving PCI,5G KPI PCell RF Serving SS-RSRP [dBm],5G KPI PCell RF Serving SS-SINR [dB],5G KPI PCell Layer2 MAC DL Throughput [Mbps],Measurement PCell Neighbor Cell Top Set(Cell Level) Top 1 PCI,Measurement PCell Neighbor Cell Top Set(Cell Level) Top 2 PCI,Measurement PCell Neighbor Cell Top Set(Cell Level) Top 3 PCI,Measurement PCell Neighbor Cell Top Set(Cell Level) Top 4 PCI,Measurement PCell Neighbor Cell Top Set(Cell Level) Top 5 PCI,Measurement PCell Neighbor Cell Top Set(Cell Level) Top 1 Filtered Tx BRSRP [dBm],Measurement PCell Neighbor Cell Top Set(Cell Level) Top 2 Filtered Tx BRSRP [dBm],Measurement PCell Neighbor Cell Top Set(Cell Level) Top 3 Filtered Tx BRSRP [dBm],Measurement PCell Neighbor Cell Top Set(Cell Level) Top 4 Filtered Tx BRSRP [dBm],Measurement PCell Neighbor Cell Top Set(Cell Level) Top 5 Filtered Tx BRSRP [dBm],5G KPI PCell Layer1 DL RB Num (Including 0)
0,2025-05-07 15:23:52.000000,128.188169,32.579273,1,712,-77.00,15.93,1351.25,258,71,284,-,-,-90.49,-95.82,-102.2,-,-,210.91
1,2025-05-07 15:23:53.000000,128.188140,32.579223,2,71,-80.97,6.60,366.57,258,712,129,-,-,-88.72,-93.26,-105.47,-,-,211.23
2,2025-05-07 15:23:54.000000,128.188117,32.579174,16,71,-85.50,1.81,334.00,258,712,-,-,-,-84.41,-96.46,-,-,-,210.99
3,2025-05-07 15:23:55.000000,128.188103,32.579113,14,71,-88.21,5.40,431.94,712,258,-,-,-,-88.04,-93.72,-,-,-,195.26
4,2025-05-07 15:23:56.000000,128.188088,32.579075,19,71,-78.45,13.59,566.34,712,258,284,-,-,-87.97,-98.70,-105.92,-,-,209.00


In [9]:
user_df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 10 entries, 0 to 9
Data columns (total 19 columns):
 #   Column                                                                             Non-Null Count  Dtype  
---  ------                                                                             --------------  -----  
 0   Timestamp                                                                          10 non-null     object 
 1   Longitude                                                                          10 non-null     float64
 2   Latitude                                                                           10 non-null     float64
 3   GPS Speed (km/h)                                                                   10 non-null     int64  
 4   5G KPI PCell RF Serving PCI                                                        10 non-null     int64  
 5   5G KPI PCell RF Serving SS-RSRP [dBm]                                              10 non-null     float64
 6

# Telecommunication User Plane Table – Column Definitions

## Summary of Telecommunication User Plane Columns

| Column | What it means | Simple intuition | Likely importance |
|----------|----------------|------------------|-------------------|
| `Timestamp` | Time of measurement | When the signal was recorded | Useful for time patterns |
| `Longitude`, `Latitude` | GPS coordinates | Where the car/device was | Important for mapping weak areas |
| `GPS Speed (km/h)` | Vehicle speed | How fast the user is moving | High speed can hurt signal stability |
| `5G KPI PCell RF Serving PCI` | Serving cell/tower ID | Which 5G tower the phone is connected to | Very important |
| `5G KPI PCell RF Serving SS-RSRP [dBm]` | Signal strength | *"How strong is the signal?"* | Very important |
| `5G KPI PCell RF Serving SS-SINR [dB]` | Signal quality/interference | *"How clean is the signal?"* | Extremely important |
| `5G KPI PCell Layer2 MAC DL Throughput [Mbps]` | Download throughput | Actual internet speed | Likely target variable |
| `Measurement PCell Neighbor Cell Top Set(Cell Level) Top 1–5 PCI` | Neighboring towers | Nearby cells competing/helping | Useful for interference and handover analysis |
| `Measurement PCell Neighbor Cell Top Set(Cell Level) Top 1–5 Filtered Tx BRSRP [dBm]` | Neighbor signal strengths | Strength of nearby towers | Important |
| `5G KPI PCell Layer1 DL RB Num (Including 0)` | Resource blocks allocated | How much radio bandwidth the user received | Extremely important |

### Quick Intuition

- **RSRP** → Signal strength  
- **SINR** → Signal quality / interference  
- **PCI** → Tower/cell identifier  
- **Neighbor PCI/BRSRP** → Nearby competing towers  
- **RB Num** → Amount of network resources assigned  
- **Throughput** → Actual internet speed experienced by the user  
- **GPS Speed** → Mobility effect on network stability

Since this dataset comes from **5G wireless network drive-test logs**, many variables represent **signal strength, signal quality, neighboring cell behavior, mobility, and network resource allocation**.  

The goal appears to be understanding **why throughput drops below 600 Mbps**, so understanding these KPIs is important before modeling.

---

## 1. Timestamp

### `Timestamp`

**Definition:**  
The time at which the network measurement was recorded.

**Business Meaning:**  
Represents when the device measured network performance during the drive test.

**Why it matters:**  
Can help identify temporal patterns or sequence-based behaviors (e.g., throughput drops after handovers).

---

## 2. Location Information

### `Longitude`  
### `Latitude`

**Definition:**  
GPS coordinates of the test vehicle.

**Business Meaning:**  
Represents the physical location where the measurement was taken.

**Why it matters:**  
Useful for identifying:

- Geographic weak coverage zones
- Areas with interference
- Specific road sections with poor throughput

Example use case:
> Throughput consistently drops near a certain location.

---

## 3. Vehicle Speed

### `GPS Speed (km/h)`

**Definition:**  
The speed of the test vehicle during the measurement.

**Business Meaning:**  
Represents how fast the user is moving through the network.

**Why it matters:**  
Higher speeds may negatively impact network performance due to:

- Frequent cell switching (handovers)
- Reduced signal stability

Typical interpretation:

| Speed | Impact |
|--------|--------|
| Low speed | Stable connection |
| High speed (>40 km/h) | Possible throughput degradation |

Related root cause:
- **C7:** Test vehicle speed exceeds 40 km/h.

---

## 4. Serving Cell Identifier

### `5G KPI PCell RF Serving PCI`

**Definition:**  
PCI (**Physical Cell Identifier**) of the serving cell.

**Business Meaning:**  
Represents which 5G cell tower (or sector) the device is currently connected to.

**Why it matters:**  

Changes in PCI indicate **handover events**, where the device switches towers.

Frequent PCI switching may indicate:

- Poor network stability
- Excessive handovers

Related root cause:
- **C5:** Frequent handovers degrade performance.

Example:

```text
712 → connected to tower A
258 → connected to tower B
71 → connected to tower C
```

---

## 5. Signal Strength

### `5G KPI PCell RF Serving SS-RSRP [dBm]`

**Definition:**  
**SS-RSRP (Synchronization Signal Reference Signal Received Power)** measures signal strength from the serving cell.

**Business Meaning:**  
Represents how strong the network signal is.

**Why it matters:**  
Weak signal strength often leads to poor throughput.

Typical interpretation:

| RSRP (dBm) | Signal Quality |
|------------|----------------|
| > -80 | Excellent |
| -80 to -90 | Good |
| -90 to -100 | Fair |
| < -100 | Weak |

Examples:

```text
-77 dBm → Strong signal
-88 dBm → Moderate signal
-105 dBm → Weak signal
```

Potential causes of weak signal:

- **C1:** Excessive downtilt angle
- **C2:** Coverage overshooting

---

## 6. Signal Quality / Interference

### `5G KPI PCell RF Serving SS-SINR [dB]`

**Definition:**  
**SS-SINR (Signal-to-Interference-plus-Noise Ratio)** measures signal quality.

**Business Meaning:**  
Represents how clean the signal is relative to surrounding interference.

**Why it matters:**  

A signal can be strong but still perform poorly if interference is high.

Typical interpretation:

| SINR (dB) | Quality |
|------------|----------|
| > 20 | Excellent |
| 13–20 | Good |
| 0–13 | Moderate |
| < 0 | Poor |

Examples:

```text
19.8 dB → Excellent signal quality
6.6 dB → Moderate quality
1.81 dB → High interference likely
```

Potential root causes:

- **C4:** Overlapping coverage interference
- **C6:** PCI interference

---

## 7. Throughput 

### `5G KPI PCell Layer2 MAC DL Throughput [Mbps]`

**Definition:**  
Actual download throughput measured in **Mbps**.

**Business Meaning:**  
Represents the real network speed experienced by the user.

**Why it matters:**  
This is likely the **target variable** for prediction or root-cause analysis.

Example:

```text
1351 Mbps → High performance
566 Mbps → Below threshold
334 Mbps → Poor performance
```

Potential binary target:

```python
df["low_throughput"] = (
    df["5G KPI PCell Layer2 MAC DL Throughput [Mbps]"] < 600
).astype(int)
```

---

## 8. Neighboring Cell PCI Information

### `Measurement PCell Neighbor Cell Top Set(Cell Level) Top 1–5 PCI`

**Definition:**  
PCI values for neighboring cells.

**Business Meaning:**  
Represents nearby cell towers that may compete with or assist the serving cell.

**Why it matters:**  

Helps identify:

- Coverage overlap
- Better neighboring cells
- Potential handover opportunities

Potential root causes:

- **C3:** Neighboring cell provides better throughput
- **C5:** Frequent handovers

---

## 9. Neighbor Signal Strength

### `Measurement PCell Neighbor Cell Top Set(Cell Level) Top 1–5 Filtered Tx BRSRP [dBm]`

**Definition:**  
Signal strength of neighboring cells.

**Business Meaning:**  
Measures how strong nearby cell signals are.

**Why it matters:**  

Can indicate whether the device is connected to the best tower.

Example:

```text
Serving Cell RSRP = -88 dBm
Neighbor Cell RSRP = -84 dBm
```

Interpretation:
> A neighboring cell has a stronger signal than the serving cell, which may indicate poor cell selection or delayed handover.

Potential causes:

- Handover inefficiency
- Coverage overshooting
- Suboptimal serving cell

---

## 10. Resource Allocation

### `5G KPI PCell Layer1 DL RB Num (Including 0)`

**Definition:**  
Number of **Resource Blocks (RBs)** allocated to the user.

**Business Meaning:**  
Represents how much radio bandwidth is assigned by the network.

**Why it matters:**  

Fewer allocated RBs may limit throughput.

Typical interpretation:

| RB Count | Impact |
|-----------|--------|
| High | Better throughput |
| Low | Reduced throughput |

Related root cause:

- **C8:** Average scheduled RBs below 160

Example:

```python
df["low_rb"] = (
    df["5G KPI PCell Layer1 DL RB Num (Including 0)"] < 160
)
```

---

# Summary of Feature Categories

### Signal Strength Features
- `SS-RSRP`
- Neighbor `BRSRP`

### Signal Quality Features
- `SS-SINR`

### Mobility Features
- `GPS Speed`
- `Serving PCI`

### Network Resource Features
- `RB Num`

### Coverage & Interference Features
- Neighbor PCI
- Neighbor Signal Strength

---

## Business Objective

The primary objective of this dataset appears to be:

> **Identify the root cause of throughput degradation (below 600 Mbps) using telecom KPIs and engineering parameters.**

This can be framed as either:

### Classification Problem
Predict:
- Root cause category (`C1–C8`)

### Root Cause Analysis
Explain:
> *Why did throughput drop in a particular road section?*

### Engineering Table Explanation

In [8]:
eng_df.head()

,gNodeB ID,Cell ID,Longitude,Latitude,Mechanical Azimuth,Mechanical Downtilt,Digital Tilt,Digital Azimuth,Beam Scenario,Height,PCI,TxRx Mode,Max Transmit Power,Antenna Model
0,34038,4,128.188273,32.579129,50,6,4,0,DEFAULT,5.0,712,64T64R,34.9,NR AAU 2
1,33164,27,128.217634,32.580720,240,3,10,0,DEFAULT,29.7,71,32T32R,34.9,NR AAU 1
2,34038,2,128.188282,32.579099,140,10,4,0,DEFAULT,5.0,258,64T64R,34.9,NR AAU 2
3,34038,24,128.187690,32.577910,40,44,9,10,SCENARIO_9,56.0,284,64T64R,34.9,NR AAU 2
4,33164,14,128.189481,32.581043,150,20,255,0,DEFAULT,28.7,129,64T64R,34.9,NR AAU 2


In [13]:
eng_df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 7 entries, 0 to 6
Data columns (total 16 columns):
 #   Column               Non-Null Count  Dtype  
---  ------               --------------  -----  
 0   gNodeB ID            7 non-null      int64  
 1   Cell ID              7 non-null      int64  
 2   Longitude            7 non-null      float64
 3   Latitude             7 non-null      float64
 4   Mechanical Azimuth   7 non-null      int64  
 5   Mechanical Downtilt  7 non-null      int64  
 6   Digital Tilt         7 non-null      int64  
 7   Digital Azimuth      7 non-null      int64  
 8   Beam Scenario        7 non-null      object 
 9   Height               7 non-null      float64
 10  PCI                  7 non-null      int64  
 11  TxRx Mode            7 non-null      object 
 12  Max Transmit Power   7 non-null      float64
 13  Antenna Model        7 non-null      object 
 14  example_id           7 non-null      int64  
 15  answer               7 non-null      object 

# Telecommunication Engineering Parameters Table – Column Definitions

The engineering parameter table contains **cell tower configuration settings** and **radio network deployment information**.

Unlike the user plane table (which captures real-time measurements), this table represents the **network configuration and physical characteristics of each cell tower**.

---

## Summary of Engineering Parameter Columns

| Column | What it means | Simple intuition | Likely importance |
|----------|----------------|------------------|-------------------|
| `gNodeB ID` | 5G base station identifier | Which physical tower/site this cell belongs to | Important |
| `Cell ID` | Sector/cell identifier within a tower | Different directional sectors on the same tower | Important |
| `Longitude`, `Latitude` | Tower GPS coordinates | Where the tower is located | Important for coverage analysis |
| `Mechanical Azimuth` | Physical antenna direction | Which direction the antenna physically faces | Very important |
| `Mechanical Downtilt` | Physical downward antenna angle | How much the antenna points downward | Extremely important |
| `Digital Tilt` | Software-controlled tilt adjustment | Fine-tuning of antenna angle | Extremely important |
| `Digital Azimuth` | Software-controlled directional adjustment | Small steering correction | Important |
| `Beam Scenario` | Antenna beam configuration | Width and shape of signal coverage | Extremely important |
| `Height` | Tower/antenna height | How high the antenna is installed | Very important |
| `PCI` | Physical Cell Identifier | Tower/cell ID connected to the device | Very important |
| `TxRx Mode` | Antenna transmission mode | Number of transmit/receive antennas | Important |
| `Max Transmit Power` | Maximum power output | Signal broadcast strength | Important |
| `Antenna Model` | Hardware type of antenna | Physical antenna equipment | Moderate importance |
| `example_id` | Example/question identifier | Which dataset example this row belongs to | Metadata |
| `answer` | Root cause label (`C1–C8`) | True diagnosis for throughput issue | Target variable |

---

# Detailed Column Explanation

## 1. gNodeB ID

### `gNodeB ID`

**Definition:**  
Unique identifier for a **5G base station (tower site)**.

**Business Meaning:**  
Represents the physical network site where the antenna is deployed.

**Why it matters:**  
Multiple cells can belong to the same tower.

Example:

```text
gNodeB 34038
├── Cell 2
├── Cell 4
└── Cell 24
```

---

## 2. Cell ID

### `Cell ID`

**Definition:**  
Unique identifier for a cell sector within a base station.

**Business Meaning:**  
Represents a directional coverage sector of a tower.

**Why it matters:**  
A single tower usually serves multiple directions.

Think:

> One tower → multiple directional cells

---

## 3. Tower Location

### `Longitude`
### `Latitude`

**Definition:**  
GPS coordinates of the cell tower.

**Business Meaning:**  
Shows where the antenna infrastructure is physically located.

**Why it matters:**  

Useful for:

- Coverage analysis
- Distance calculations
- Overshooting detection

Related root cause:

- **C2:** Coverage distance exceeds 1 km

---

## 4. Mechanical Azimuth

### `Mechanical Azimuth`

**Definition:**  
Physical direction (angle) the antenna faces.

**Business Meaning:**  
Represents where the antenna is pointed horizontally.

Measured in degrees:

| Degree | Direction |
|---------|------------|
| 0° | North |
| 90° | East |
| 180° | South |
| 270° | West |

Example:

```text
50° → Northeast
140° → Southeast
240° → Southwest
```

**Why it matters:**  

Poor antenna direction may lead to:

- Weak coverage
- Overshooting
- Wrong serving cell

---

## 5. Mechanical Downtilt

### `Mechanical Downtilt`

**Definition:**  
Physical downward tilt of the antenna.

**Business Meaning:**  
Controls how much the signal points downward.

**Why it matters:**  

One of the most important telecom variables.

Too much downtilt:

> Signal dies too early

Too little downtilt:

> Signal travels too far (overshooting)

Related root causes:

- **C1:** Downtilt too large
- **C2:** Coverage overshooting

Example:

```text
3° → broad coverage
20° → aggressive downward tilt
44° → extremely tilted
```

---

## 6. Digital Tilt

### `Digital Tilt`

**Definition:**  
Software-controlled tilt adjustment.

**Business Meaning:**  
Fine-tunes antenna coverage without physically moving hardware.

**Special Rule:**

```text
255 = default 6°
```

Other values represent actual tilt degrees.

Example:

```text
255 → 6°
10 → 10°
4 → 4°
```

**Why it matters:**  

Can strongly influence:

- Signal strength
- Coverage distance
- Throughput

Related root causes:

- **C1**
- **C2**

---

## 7. Digital Azimuth

### `Digital Azimuth`

**Definition:**  
Software steering of antenna direction.

**Business Meaning:**  
Allows slight directional adjustments.

**Why it matters:**  

Can affect:

- Signal overlap
- Neighbor interference
- Coverage optimization

---

## 8. Beam Scenario

### `Beam Scenario`

**Definition:**  
Determines vertical beamwidth of the antenna.

**Business Meaning:**  
Controls how wide or narrow coverage beams are.

Beamwidth mapping:

| Scenario | Vertical Beamwidth |
|-----------|--------------------|
| DEFAULT / SCENARIO_1–5 | 6° |
| SCENARIO_6–11 | 12° |
| SCENARIO_12+ | 25° |

**Why it matters:**  

Narrow beams:
> Better focus, longer range

Wide beams:
> Wider coverage, more overlap

Potential causes:

- Coverage overlap (**C4**)
- Coverage overshooting (**C2**)

---

## 9. Height

### `Height`

**Definition:**  
Installation height of antenna.

**Business Meaning:**  
How high the tower antenna sits above ground.

**Why it matters:**  

Higher antennas may:

- Increase coverage distance
- Cause overshooting

Related root cause:

- **C2**

Example:

```text
5m → low deployment
56m → tall tower
123m → very high
```

---

## 10. PCI

### `PCI`

**Definition:**  
**Physical Cell Identifier**.

**Business Meaning:**  
Unique identifier of the serving cell.

**Why it matters:**  

Used to determine:

> Which tower the user device connects to

Related root cause:

- **C6:** Same PCI mod 30 interference

---

## 11. TxRx Mode

### `TxRx Mode`

**Definition:**  
Transmission/Reception antenna configuration.

**Business Meaning:**  
Number of transmit and receive antennas.

Examples:

| Mode | Meaning |
|-------|----------|
| `4T4R` | 4 transmit / 4 receive |
| `32T32R` | 32 transmit / 32 receive |
| `64T64R` | 64 transmit / 64 receive |

**Why it matters:**  

More antennas generally improve:

- Capacity
- Beamforming
- Throughput

---

## 12. Max Transmit Power

### `Max Transmit Power`

**Definition:**  
Maximum signal power emitted by the antenna.

**Business Meaning:**  
How strongly the tower broadcasts signal.

**Why it matters:**  

Higher power may:

- Increase coverage
- Improve signal strength
- Cause overlap/interference

---

## 13. Antenna Model

### `Antenna Model`

**Definition:**  
Physical hardware model of antenna.

**Business Meaning:**  
Represents the deployed equipment.

Examples:

```text
NR AAU 1
NR AAU 2
NR AAU 3
```

**Why it matters:**  

Different models may support:

- Different beamforming capability
- Different throughput capacity

---

## 14. example_id

### `example_id`

**Definition:**  
Unique identifier for each dataset example.

**Business Meaning:**  
Used to link records back to the original question/example.

---

## 15. answer

### `answer`

**Definition:**  
True root cause label.

**Business Meaning:**  
Represents the diagnosed reason for throughput degradation.

Classes:

| Root Cause | Meaning |
|-------------|----------|
| `C1` | Downtilt too large |
| `C2` | Coverage overshooting |
| `C3` | Neighboring cell better |
| `C4` | Overlapping coverage |
| `C5` | Frequent handovers |
| `C6` | PCI interference |
| `C7` | Vehicle speed issue |
| `C8` | Low resource blocks |

This is the **target variable** for machine learning.

# Questions about the data
1. Should we merge User Plane Table and Engineering Table Explanation? -> 
2. Each one data row has multiple rows of userplane and enginiring engineering table.

In [10]:
parsed_rows = []

for i, row in df.iterrows():
    user_df, eng_df = parse_question_tables(row["question"])

    user_df["example_id"] = i
    eng_df["example_id"] = i
    user_df["answer"] = row["answer"]
    eng_df["answer"] = row["answer"]

    parsed_rows.append((user_df, eng_df))

all_user_data = pd.concat([x[0] for x in parsed_rows], ignore_index=True)
all_engineering_data = pd.concat([x[1] for x in parsed_rows], ignore_index=True)

In [12]:
all_user_data

,Timestamp,Longitude,Latitude,GPS Speed (km/h),5G KPI PCell RF Serving PCI,5G KPI PCell RF Serving SS-RSRP [dBm],5G KPI PCell RF Serving SS-SINR [dB],5G KPI PCell Layer2 MAC DL Throughput [Mbps],Measurement PCell Neighbor Cell Top Set(Cell Level) Top 1 PCI,Measurement PCell Neighbor Cell Top Set(Cell Level) Top 2 PCI,...,Measurement PCell Neighbor Cell Top Set(Cell Level) Top 4 PCI,Measurement PCell Neighbor Cell Top Set(Cell Level) Top 5 PCI,Measurement PCell Neighbor Cell Top Set(Cell Level) Top 1 Filtered Tx BRSRP [dBm],Measurement PCell Neighbor Cell Top Set(Cell Level) Top 2 Filtered Tx BRSRP [dBm],Measurement PCell Neighbor Cell Top Set(Cell Level) Top 3 Filtered Tx BRSRP [dBm],Measurement PCell Neighbor Cell Top Set(Cell Level) Top 4 Filtered Tx BRSRP [dBm],Measurement PCell Neighbor Cell Top Set(Cell Level) Top 5 Filtered Tx BRSRP [dBm],5G KPI PCell Layer1 DL RB Num (Including 0),example_id,answer
0,2025-05-07 15:23:52.000000,128.188169,32.579273,1,712,-77.00,15.93,1351.25,258,71,...,-,-,-90.49,-95.82,-102.2,-,-,210.91,0,C2
1,2025-05-07 15:23:53.000000,128.188140,32.579223,2,71,-80.97,6.60,366.57,258,712,...,-,-,-88.72,-93.26,-105.47,-,-,211.23,0,C2
2,2025-05-07 15:23:54.000000,128.188117,32.579174,16,71,-85.50,1.81,334.00,258,712,...,-,-,-84.41,-96.46,-,-,-,210.99,0,C2
3,2025-05-07 15:23:55.000000,128.188103,32.579113,14,71,-88.21,5.40,431.94,712,258,...,-,-,-88.04,-93.72,-,-,-,195.26,0,C2
4,2025-05-07 15:23:56.000000,128.188088,32.579075,19,71,-78.45,13.59,566.34,712,258,...,-,-,-87.97,-98.70,-105.92,-,-,209.00,0,C2
5,2025-05-07 15:23:57.000000,128.188073,32.579033,22,258,-77.45,11.45,1308.74,712,71,...,129,-,-89.56,-93.31,-104.18,-110.78,-,204.95,0,C2
6,2025-05-07 15:23:58.000000,128.188051,32.579002,16,258,-81.66,10.06,1310.14,712,71,...,129,-,-84.85,-93.30,-102.87,-114.09,-,210.13,0,C2
7,2025-05-07 15:23:59.000000,128.188073,32.578995,26,258,-79.13,16.32,1023.36,712,71,...,129,-,-86.75,-93.71,-104.01,-107.26,-,208.06,0,C2
8,2025-05-07 15:24:00.000000,128.188036,32.578984,33,258,-74.09,19.80,770.83,712,71,...,129,-,-89.29,-96.40,-100.21,-114.11,-,207.07,0,C2
9,2025-05-07 15:24:01.000000,128.188036,32.578984,25,258,-71.09,19.62,1014.75,712,71,...,-,-,-90.14,-92.93,-,-,-,205.57,0,C2


In [11]:
all_engineering_data

,gNodeB ID,Cell ID,Longitude,Latitude,Mechanical Azimuth,Mechanical Downtilt,Digital Tilt,Digital Azimuth,Beam Scenario,Height,PCI,TxRx Mode,Max Transmit Power,Antenna Model,example_id,answer
0,34038,4,128.188273,32.579129,50,6,4,0,DEFAULT,5.0,712,64T64R,34.9,NR AAU 2,0,C2
1,33164,27,128.217634,32.580720,240,3,10,0,DEFAULT,29.7,71,32T32R,34.9,NR AAU 1,0,C2
2,34038,2,128.188282,32.579099,140,10,4,0,DEFAULT,5.0,258,64T64R,34.9,NR AAU 2,0,C2
3,34038,24,128.187690,32.577910,40,44,9,10,SCENARIO_9,56.0,284,64T64R,34.9,NR AAU 2,0,C2
4,33164,14,128.189481,32.581043,150,20,255,0,DEFAULT,28.7,129,64T64R,34.9,NR AAU 2,0,C2
5,33164,14,128.189202,32.581003,150,20,255,0,DEFAULT,28.7,129,64T64R,34.9,NR AAU 2,1,C1
6,33164,24,128.188958,32.581776,235,15,9,0,DEFAULT,22.1,368,32T32R,34.9,NR AAU 1,1,C1
7,33164,27,128.189155,32.580940,260,44,9,0,DEFAULT,29.7,71,32T32R,34.9,NR AAU 1,1,C1
8,34036,14,128.200652,32.594683,0,0,255,0,EXPAND_SCENARIO_2,0.0,868,4T4R,46.0,Other,1,C1
9,33164,12,128.190042,32.583484,45,6,4,0,DEFAULT,22.0,151,64T64R,34.9,NR AAU 2,1,C1


| Root Cause                    | Important Features                               |
| ----------------------------- | ------------------------------------------------ |
| C1 Downtilt too large         | total tilt, beamwidth, weak far-end signal       |
| C2 Overshooting (>1km)        | serving distance, low downtilt, far serving cell |
| C3 Neighbor higher throughput | neighbor stronger than serving                   |
| C4 Overlapping coverage       | multiple strong neighbors                        |
| C5 Frequent handovers         | handover count / PCI changes                     |
| C6 PCI mod 30 interference    | serving PCI % 30 == neighbor PCI % 30            |
| C7 Speed issue                | average/max speed > 40                           |
| C8 Low RB scheduling          | avg RB < 160                                     |



